modules/standard/Map.chpl

```diff
@@ -419,26 +419,26 @@ module Map {
         var val: valType;
         table.fillSlot(slot, k, val);
       }
-      return table.table[slot].val;
+      ref result = table.table[slot].val;
+      return result;
     }
 
-    pragma "no doc"
-    proc const this(k: keyType) const throws
-      where shouldReturnRvalueByValue(valType) &&
-            !isNonNilableClass(valType) {
+    proc ref this(k: keyType) ref throws
+      where isNonNilableClass(valType) {
```

Comment:
I wonder if the change in behavior for that future has to do with this change, since the map there contains not a `nonNilableClass`, but a record containing a `nonNilableClass` (whose default initializer doesn't initialize it.), which presumably should be handled the same?

I also don't follow `shouldReturnRvalueByValue()`, but is the concern of this where clause the valtype's default-initializability, rather than it specifically being a non-nilable class?  (Just skimming, why does this need the where clause at all, since it'll throw in the case where it would need a new object?)  (Feel free to ignore me if I'm off in the weeds barking up the wrong tree.)


Semi-relatedly,  as I'm looking at this, the chpldoc comment for this set of functions says

```
Get the value mapped to the given key, or add the mapping if key does not
      exist.
```

and

```
      :throws: `KeyNotFoundError` if `k` not in map
```

which contradict each other.  (Though I don't have any suggestions to clarify that succinctly.)

Reply:
> why does this need the where clause at all, since it'll throw in the case where it would need a new object?

This is a good point, I think removing the `where` clause is the right thing to do, so all non-default initializable values will throw if their value is not found in the map (which is what the old `getReference` did as well).

I have also updated the doc string to be more accurate! Thanks for the review.